File: BinaryTree/python/treeLevels.py
from collections import deque
from typing import List, Optional

from node import Node


def tree_levels(root: Optional[Node]) -> List[List[int]]:
    if root is None:
        return []

    levels: List[List[int]] = []
    queue = deque([(root, 0)])
    while queue:
        current, level = queue.popleft()
        if len(levels) == level:
            levels.append([])
        levels[level].append(current.val)
        if current.left:
            queue.append((current.left, level + 1))
        if current.right:
            queue.append((current.right, level + 1))
    return levels

File: BinaryTree/csharp/TreeLevels.cs
using System.Collections.Generic;

namespace BinaryTreeSolutions;

public static class TreeLevels
{
    public static IList<IList<int>> Solve(TreeNode? root)
    {
        var levels = new List<IList<int>>();
        if (root is null)
        {
            return levels;
        }

        var queue = new Queue<(TreeNode Node, int Level)>();
        queue.Enqueue((root, 0));
        while (queue.Count > 0)
        {
            var (node, level) = queue.Dequeue();
            if (levels.Count == level)
            {
                levels.Add(new List<int>());
            }
            levels[level].Add(node.Val);
            if (node.Left is not null) queue.Enqueue((node.Left, level + 1));
            if (node.Right is not null) queue.Enqueue((node.Right, level + 1));
        }

        return levels;
    }
}
